service/src/main/java/com/epam/esm/service/impl/CertificateServiceImpl.java

```diff
@@ -0,0 +1,127 @@
+package com.epam.esm.service.impl;
+
+import com.epam.esm.dto.CertificateDto;
+import com.epam.esm.dto.mapper.CertificateMapper;
+import com.epam.esm.entity.Certificate;
+import com.epam.esm.exception.EntityAlreadyExistsException;
+import com.epam.esm.exception.EntityNotFoundException;
+import com.epam.esm.exception.InvalidEntityException;
+import com.epam.esm.repository.SearchCriteria;
+import com.epam.esm.repository.repositoryinterfaces.CertificateRepository;
+import com.epam.esm.repository.repositoryinterfaces.TagRepository;
+import com.epam.esm.service.CertificateService;
+import com.epam.esm.validator.CertificateValidator;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+@Service
+public class CertificateServiceImpl implements CertificateService {
+
+    private final CertificateRepository certificateRepository;
+    private final TagRepository tagRepository;
+    private final CertificateMapper certificateMapper;
+    private final CertificateValidator certificateValidator;
+
+    public CertificateServiceImpl(CertificateRepository certificateRepository,
+                                  TagRepository tagRepository,
+                                  @Autowired CertificateMapper certificateMapper,
+                                  @Autowired CertificateValidator certificateValidator) {
+        this.certificateRepository = certificateRepository;
+        this.tagRepository = tagRepository;
+        this.certificateMapper = certificateMapper;
+        this.certificateValidator = certificateValidator;
+    }
+
+    @Override
+    public List<CertificateDto> findCertificateByCriteriaService(SearchCriteria searchCriteria) {
+
+        if (certificateValidator.areParamsEmpty(searchCriteria.getTagName(), searchCriteria.getCertificateName()
```

Comment:
Why don't create SearchCriteriaValidator and pass searchCriteria as single argument?